tree-wide: add support for connecting to AF_UNIX sockets in the file system beyond the 108ch limit

poettering opened the pull request:
This adds some infrastructure to make sure we can connect to AF_UNIX sockets in the filesystem beyond the 108 char limit that `sockaddr_un` enforces. It uses O_PATH to work around the limit (i.e. connects to `/proc/self/fd/<fd>` instead of the literal path).

Most of our internal sockets do not use such paths, but there's value to allowing paths configured externally to be connectible regardless of their length.

yuwata commented on src/basic/socket-util.c:
```diff
@@ -1426,3 +1426,51 @@ int socket_get_mtu(int fd, int af, size_t *ret) {
+                if (inode_fd < 0)
+                        return -errno;
+
+                proc = FORMAT_PROC_FD_PATH(inode_fd);
```
Is this safe?? I thought it can be only used as a function argument.

poettering replied:
this is safe, c99 compound literals are valid from the point on they are declared. What's key though is that the memory is only valid within the current {} block. Which we comply here by, we copy it into the sockaddr_un further down, via memcpy(), and then don't need the memory anmyore.

poettering replied:
this is safe, c99 compound literals are valid from the point on they are declared. What's key though is that the memory is only valid within the current {} block. Which we comply here by, we copy it into the sockaddr_un further down, via memcpy(), and then don't need the memory anmyore.

yuwata commented on src/shared/nscd-flush.c:
```diff
@@ -1,4 +1,5 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
+#include <fcntl.h>
```
Not a matter, but we usually add a blank line after SPDX header.

bluca commented:
```
../src/basic/fileio.c:766:37: error: unused variable 'dfd' [-Werror,-Wunused-variable]
                _cleanup_close_ int dfd = -1, sk = -1;
                                    ^
```

floppym commented on src/basic/socket-util.c:
```diff
@@ -1426,3 +1426,51 @@ int socket_get_mtu(int fd, int af, size_t *ret) {
+
+                /* if dir_fd is specified, then we need to go the indirect O_PATH route, because connectat()
+                 * does not exist. If the path is too long, we also need to take the indirect route, since we
+                 * cant fit this into a sockaddr_run directly. */
```
Typo: `sockaddr_run` should be `sockaddr_un`.

bluca replied:
also /cant/can't/

poettering commented:
> So the path will be invalid after the function returns, right? IE, `/proc/self/fd/<fd>` will be closed and no longer point to the socket once the connection completes. Would that cause issues? Especially given it could be reused at some point? Also other programs will see `/proc/self/fd<fd>` eg, in `netstat -x`, but obviously `self` is actually another process?

When connecting to a socket the `/proc/self/fd/<fd>` should note have any negative effects I would be aware of. Things like /proc/net/unix don't show the target path used for connection anyway... AF_UNIX sockets in the file system are weird beasts, they only act as entrypoints, and the connection socket does not carry info about the entrypoint anymore. i.e. if you do getpeername() of an AF_UNIX socket you connected like that you'll not get info about the socket node in the fs that you connected to, but about the AF_UNIX inode that is behind it...

bluca commented:
> > So the path will be invalid after the function returns, right? IE, `/proc/self/fd/<fd>` will be closed and no longer point to the socket once the connection completes. Would that cause issues? Especially given it could be reused at some point? Also other programs will see `/proc/self/fd<fd>` eg, in `netstat -x`, but obviously `self` is actually another process?
> 
> When connecting to a socket the `/proc/self/fd/<fd>` should note have any negative effects I would be aware of. Things like /proc/net/unix don't show the target path used for connection anyway... AF_UNIX sockets in the file system are weird beasts, they only act as entrypoints, and the connection socket does not carry info about the entrypoint anymore. i.e. if you do getpeername() of an AF_UNIX socket you connected like that you'll not get info about the socket node in the fs that you connected to, but about the AF_UNIX inode that is behind it...

Got it, thanks

poettering commented:
> When connecting to a socket the /proc/self/fd/<fd> should note have any negative effects I would be aware of. Things like /proc/net/unix don't show the target path used for connection anyway... AF_UNIX sockets in the file system are weird beasts, they only act as entrypoints, and the connection socket does not carry info about the entrypoint anymore. i.e. if you do getpeername() of an AF_UNIX socket you connected like that you'll not get info about the socket node in the fs that you connected to, but about the AF_UNIX inode that is behind it...

Also, what's the alternative here? that is to simply fail. I think making this just work is the better approach even if it's a rather arcane indirection we have to implement here...

poettering commented:
New version pushed, addressing all issues raised. ptal. given this fixes a bad error propagation and comes with tests (i.e. should be relatively safe) I think this should probably be merged before 251. but i could also just cherry-pick that one relevant patch...

yuwata commented on src/userdb/userdbctl.c:
```diff
@@ -942,25 +942,19 @@ static int display_services(int argc, char *argv[], void *userdata) {
-
                 fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
                 if (fd < 0)
                         return log_error_errno(r, "Failed to allocate AF_UNIX/SOCK_STREAM socket: %m");
```
Another bug. r -> errno.

yuwata commented:
Could you add a test for connect_unix_path()?

bluca commented:
> New version pushed, addressing all issues raised. ptal. given this fixes a bad error propagation and comes with tests (i.e. should be relatively safe) I think this should probably be merged before 251. but i could also just cherry-pick that one relevant patch...

It's very late in the cycle, I think it would safer to just cherry-pick the error propagation fix and merge the rest later? It would probably be fine, but I don't think there's an urgent need either, right?

poettering commented:
> Could you add a test for connect_unix_path()?

There's a test for this already in place, via the read_file_full() test, which tests behaviour with overlay long sockets paths. I wouldn't know which corner cases wouldn't already be covered by that...

yuwata commented:
> > New version pushed, addressing all issues raised. ptal. given this fixes a bad error propagation and comes with tests (i.e. should be relatively safe) I think this should probably be merged before 251. but i could also just cherry-pick that one relevant patch...
> 
> It's very late in the cycle, I think it would safer to just cherry-pick the error propagation fix and merge the rest later? It would probably be fine, but I don't think there's an urgent need either, right?

#23369.

> > Could you add a test for connect_unix_path()?
> 
> There's a test for this already in place, via the read_file_full() test, which tests behaviour with overlay long sockets paths. I wouldn't know which corner cases wouldn't already be covered by that...

Ah, thanks.

yuwata commented:
Rebased as #23369 is merged.
